pertpy/tools/_distances.py

```diff
@@ -0,0 +1,230 @@
+- An interface / API / wrapper class "Distance" which is accessible to the user
+and selects the appropriate distance metric class based on the user input.
+TODO:
+- Inherit docstrings from parent classes
+- Store precomputed distances in adata object
+- Implement MMD
+- Implement Wasserstein distance
+"""
+
+
+# This is the API / Interface class that the user will use
+class Distance:
+    """Distance class.
+
+    Used to compute distances between groups of cells. The distance metric can
+    be specified by the user. The class also provides a method to compute
+    the pairwise distances between all groups of cells.
+
+    Attributes:
+        metric_fct (AbstractDistance): Distance metric function.
+        obsm_key (str): Name of embedding in adata.obsm to use.
+        metric (str): Name of distance metric.
+    """
+
+    def __init__(
+        self,
+        metric: str = "edistance",
+        obsm_key: str = "X_pca",
+    ) -> None:
+        """Initialize Distance class.
+
+        Args:
+            metric (str, optional): Distance metric to use. Defaults to edistance.
+            obsm_key (str, optional): Name of embedding in adata.obsm to use.
+        """
+        metric_fct: AbstractDistance = None
+        if metric == "edistance":
+            metric_fct = Edistance()
+        elif metric == "pseudobulk":
+            metric_fct = PseudobulkDistance()
+        elif metric == "mean_pairwise":
+            metric_fct = MeanPairwiseDistance()
+        else:
+            raise ValueError(f"Metric {metric} not recognized.")
+        self.metric_fct = metric_fct
+        self.obsm_key = obsm_key
+        self.metric = metric
+
+    def __call__(
+        self,
+        X: np.ndarray,
+        Y: np.ndarray,
+        **kwargs,
+    ) -> float:
+        """Compute distance between vectors X and Y.
+
+        Args:
+            X (np.ndarray): First vector of shape (n_samples, n_features).
+            Y (np.ndarray): Second vector of shape (n_samples, n_features).
+
+        Returns:
+            float: Distance between X and Y.
+        """
+        return self.metric_fct(X, Y, **kwargs)
+
+    def pairwise(
+        self,
+        adata: AnnData,
+        groupby: str,
+        groups: list[str] | None = None,
+        verbose: bool = True,
+        **kwargs,
+    ) -> pd.DataFrame:
+        """Get pairwise distances between groups of cells.
+
+        Args:
+            adata (AnnData): Annotated data matrix.
+            groupby (str): Column name in adata.obs.
+            groups (Optional[List[str]], optional): List of groups to compute
+            pairwise distances for. Defaults to None (all groups).
+            verbose (bool, optional): Whether to show progress bar. Defaults to
+            True.
+
+        Returns:
+            pd.DataFrame: Dataframe with pairwise distances.
+        """
+        # TODO: This could make use of the precomputed distances if we store
```

Comment:
Can we make this happen?

Reply:
I made this happen.
Though this currently includes computing all pairwise distances between all cells. Scales bad with number of cells. But actually improving this might be difficult, so maybe let's keep this as a "do this someday" task.